### PR TITLE
Run in the Git root directory

### PR DESCRIPTION
**Pull request by kevgo**
@charlierudolph @allewun 

This implements #371 

This is a "best effort" implementation for the case that the user runs "git sync" in a folder that doesn't exist on main. 

The previous implementation crashes, and is unable to continue or abort, thereby at best leaving the user in a very confusing state (in a non-existing directory, without any Git functionality), and at worst  losing stashed away changes. 

With these changes, the Git Town command at least finishes properly. The user shell is still in limbo, but the user can `cd` back into his directory manually. CD'ing in the script doesn't help because it doesn't change the parent shell. 

Unfortunately I can't test this. When you are in a folder that doesn't exist on main, and Git changes to the main branch, it physically removes that folder from the hard drive. This leaves any parent process in limbo, be it your Bash/Fish shell or the surrounding Cucumber tester. Further Git or file operations are impossible. As a second-best solution, I have documented the issue in the code base to motivate and justify the respective changes.

This is still a good change imho, since it improves the reliability and robustness of Git Town in edge cases, and in general is the right thing to do. 

Steps to reproduce the issue:
* git hack test
* mkdir foo ; cd foo ; touch bar
* git add -A ; git commit -m "test"
* git sync

### CODE REVIEW

**Comment by allewun** [comment]
Would it be viable to test that this situation doesn't happen anymore? I think you can detect this situation by checking the exit code of `ls $(pwd)`

File: src/helpers/helpers.sh [hunk 1]
```diff
@@ -36,7 +36,8 @@ source "$current_dir/environment.sh" "$@"
 source "$current_dir/configuration.sh" "$@"
 
 if [ "$(is_git_repository)" == true ]; then
-  export INITIAL_BRANCH_NAME=$(get_current_branch_name)
-  export INITIAL_OPEN_CHANGES=$(has_open_changes)
-  export HAS_REMOTE=$(has_remote_url)
+export INITIAL_BRANCH_NAME=$(get_current_branch_name)
+export INITIAL_OPEN_CHANGES=$(has_open_changes)
+export HAS_REMOTE=$(has_remote_url)
+export INITIAL_DIRECTORY=$(pwd)
```
Comment (charlierudolph) [inline src/helpers/helpers.sh]: Please re-indent these or explain why you un-indented them

File: src/helpers/script_helpers.sh [hunk 2]
```diff
@@ -109,9 +109,15 @@ function run {
     undo_command
   else
     remove_step_files
+
+    # Running inside a subfolder can cause issues if that folder doesn't exist
+    # on the main branch. Hence we cd into the main branch here.
```
Comment (charlierudolph) [inline src/helpers/script_helpers.sh]: Comment doesn't make sense to me. I don't believe it has anything to do with the main branch and just depends if the folder is known to git yet or not. And you aren't "cd"ing into the main branch but into the root git folder

File: src/helpers/helpers.sh (hunk 1 above)
Comment (charlierudolph) [inline src/helpers/helpers.sh]: Also please put in alpha order

Comment (kevgo) [inline src/helpers/helpers.sh]: great points. Fixed

File: src/helpers/script_helpers.sh (hunk 2 above)
Comment (kevgo) [inline src/helpers/script_helpers.sh]: Fixed the `cd` phrasing. The problem is indeed related to the folder not existing on the main branch. If it exists in both branches, no problem.

**Comment by kevgo** [comment]
I figured out a way to spec this. Spec added.

**Comment by kevgo** [comment]
Spec for `git sync --all` added as well.

**Comment by kevgo** [comment]
all comments addressed

**Comment by charlierudolph** [comment]
I don't see the need for the `git sync --all` test. I also think this could be a single feature test. Whats your reasoning for having three separate specs?

**Comment by kevgo** [comment]
The reason for the `git sync --all` test is mostly that one could run `git sync --all` in a folder that only exists on a particular feature branch, and we want to be sure that this works similarly well to how a normal `git sync` works. If you feel strongly against this, then I can take it out, since I don't feel strongly about keeping it in.  :)

The reason for 3 separate tests is to match the existing test structure. Do you mean this could all go into one file? Where would this file go? Directly under `features/git-sync`?

**Comment by charlierudolph** [comment]
Okay. I still don't think we need to differentiate between with open changes and without open changes as thats not important to this test. I'm okay with two tests and think they should live at
```
features/git-sync/all_branches/folder_doesnt_exist_on_main.feature
features/git-sync/current_branch/feature_branch/folder_doesnt_exist_on_main.feature
```
Also I'd prefer the file name to not "doesnt" and instead of "does not" and to say "main branch" instead of just "main"
```
folder_does_not_exist_on_main_branch.feature
```

**Comment by kevgo** [comment]
You got it. All comments addressed

**Comment by kevgo** [comment]
Ping

File: src/helpers/script_helpers.sh [hunk 3]
```diff
@@ -109,9 +109,15 @@ function run {
...
     remove_step_files
+
+    # Running inside a subfolder can cause issues if that folder doesn't exist
+    # on the main branch. Hence we cd into the root directory of the Git repo here.
+    # See https://github.com/Originate/git-town/issues/371 for more details.
+    cd_git_root
     preconditions "$@"
     steps > "$STEPS_FILE"
     run_steps "$STEPS_FILE" undoable
+    cd_start_dir
```
Comment (charlierudolph) [inline src/helpers/script_helpers.sh]: What about when we hit an error and then either need to abort and continue. I think we need to add the cd's to the steps list instead.

Comment (kevgo) [inline src/helpers/script_helpers.sh]: Great point. I have refactored the solution to make cd'ing into the root directory a real, undoable, logged step. This not only displays to the user that Git Town was cd'ing out of the current directory, it also properly goes back to the directory in all situations. Specs for all cases added.

**Comment by kevgo** [comment]
All comments addressed

File: src/helpers/file_helpers.sh [hunk 4]
```diff
@@ -3,15 +3,6 @@
 # Helper methods for dealing with files and temp files.
 
 
-# Unique string that identifies the current directory
-temp_filename_suffix="$(pwd | tr '/' '_')"
```
Comment (kevgo) [inline src/helpers/file_helpers.sh]: These changes were necessary here to make the specs pass, so I couldn't extract them into a separate PR.

File: src/helpers/folder_helpers.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
+
+# Helper methods for dealing with directories.
+
+
+# Changes into the given directory
+function change_directory {
+  local directory=$1
+  pwd
+  run_git_command "cd $directory"
```
Comment (charlierudolph) [inline src/helpers/folder_helpers.sh]: This is not a git command

File: src/helpers/folder_helpers.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  pwd
```
Comment (charlierudolph) [inline src/helpers/folder_helpers.sh]: Why do you print the working directory twice?

Comment (charlierudolph) [inline src/helpers/folder_helpers.sh]: I don't think we need to print the working directory at all

File: src/helpers/helpers.sh [hunk 7]
```diff
@@ -36,7 +37,12 @@ source "$current_dir/environment.sh" "$@"
 source "$current_dir/configuration.sh" "$@"
 
 if [ "$(is_git_repository)" == true ]; then
+  temp_filename_suffix="$(git rev-parse --show-toplevel | tr '/' '_')"
```
Comment (charlierudolph) [inline src/helpers/helpers.sh]:
```sh
temp_filename_suffix=$(git_root | tr '/' '_')
```

File: src/helpers/folder_helpers.sh [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+
+function undo_steps_for_change_directory {
+  echo "change_directory $(pwd)"
+}
+
+
+# Returns whether the current working directory
+# is the root directory of the current Git workspace.
+function is_in_git_sub_directory {
```
Comment (charlierudolph) [inline src/helpers/folder_helpers.sh]: Based on the comment this should be named is in `is_in_git_root_directory`. To make use of that may need to implement `echo_if_false`. I think i'd prefer that.

Comment (charlierudolph) [inline src/helpers/folder_helpers.sh]: Nevermind fine as is but update the comment to something like
```
# Returns whether the current working directory
# is in a subdirectory of the current Git workspace.

File: features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/conflict.feature [hunk 9]
```diff
@@ -0,0 +1,87 @@
+Feature: git sync: syncing inside a folder that doesn't exist on the main branch (with open changes)
+
+  As a developer syncing inside a committed folder that doesn't exist on the main branch
+  I want the command to at least finish properly
```
Comment (charlierudolph) [inline features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/conflict.feature]: I'd say remove "at least". I want the command to finish properly

File: features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/no_conflict.feature [hunk 10]
```diff
@@ -0,0 +1,65 @@
+Feature: git sync: syncing inside a folder that doesn't exist on the main branch (with open changes)
+
+  As a developer syncing inside a committed folder that doesn't exist on the main branch
+  I want the command to at least finish properly
```
Comment (charlierudolph) [inline features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/no_conflict.feature]: remove "at least"

File: features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/conflict.feature (hunk 9 above)
Comment (charlierudolph) [inline features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/conflict.feature]: This should actually point to the user value in no_conflict.feature

File: features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/no_conflict.feature [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  Background:
+    Given I have feature branches named "current_feature" and "other_feature"
+    And the following commits exist in my repository
+      | BRANCH          | LOCATION         | MESSAGE              | FILE NAME        |
+      | main            | local and remote | main commit          | main_file        |
+      | current_feature | local and remote | folder commit        | new_folder/file1 |
+      | other_feature   | local and remote | other feature commit | file2            |
+    And I am on the "current_feature" branch
+    And I have an uncommitted file
+    When I run `git sync --all` in the "new_folder" folder
```
Comment (charlierudolph) [inline features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/no_conflict.feature]: Why is this `git sync --all`. it should be `git sync` (based on the location of the feature), which means the `other_feature` branch can be removed and `current_feature` branch can be renamed to `feature`

File: features/step_definitions/folder_steps.rb [hunk 12]
```diff
@@ -22,6 +22,16 @@
 
 
 
+Then(/^I am in the project root folder$/) do
+  expect(Dir.pwd).to eql git_root_folder
+end
+
+
+Then(/^I am in the "(.+?)" folder$/) do |folder_name|
+  expect(Dir.pwd.sub git_root_folder, '[project root]').to eql "[project root]/#{folder_name}"
```
Comment (charlierudolph) [inline features/step_definitions/folder_steps.rb]: Another way to consider:
```ruby
relativeProjectPath = Pathname.new(Dir.pwd).relative_path_from Pathname.new(git_root_folder)
expect(relativeProjectPath).to.eql folder_name
```

File: features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/conflict.feature (hunk 9 above)
Comment (kevgo) [inline features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/conflict.feature]: Both implemented

File: src/helpers/folder_helpers.sh (hunk 5 above)
Comment (kevgo) [inline src/helpers/folder_helpers.sh]: Yeah, we have some technical drift here. I have opened #493 to address this separately, because it will result in a larger shotgun refactor.

I'd like to keep printing this command to the user, and verifying it in the commands table in Cucumber, because it is a relevant operation that both the user and the feature specs should be aware of.

File: src/helpers/folder_helpers.sh (hunk 6 above)
Comment (kevgo) [inline src/helpers/folder_helpers.sh]: Sorry, debugging leftovers. This was pretty tricky to test, because of disappearing folders in the middle of the spec, which left Ruby dumbfounded.

File: src/helpers/helpers.sh (hunk 7 above)
Comment (kevgo) [inline src/helpers/helpers.sh]: Good catch. Fixed

File: src/helpers/folder_helpers.sh (hunk 8 above)
Comment (kevgo) [inline src/helpers/folder_helpers.sh]: Done

File: src/helpers/folder_helpers.sh (hunk 5 above)
Comment (charlierudolph) [inline src/helpers/folder_helpers.sh]: I don't think there is technical drift. We already have a function (`run_command`) that can be used here. Just need to update our tests so when it captures what commands are run the branch is optional.

File: features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/no_conflict.feature (hunk 11 above)
Comment (kevgo) [inline features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/no_conflict.feature]: True. I created separate specs for `git sync` and `git sync --all` now.

File: features/step_definitions/folder_steps.rb (hunk 12 above)
Comment (kevgo) [inline features/step_definitions/folder_steps.rb]: Your version is less hacky, so I've adopted it. Good stuff!

**Comment by kevgo** [comment]
All comments addressed

**Comment by charlierudolph** [comment]
I don't think the `git sync --all` test is necessary. Its not sufficiently different in my mind from the `git sync` test

File: src/helpers/folder_helpers.sh (hunk 5 above)
Comment (charlierudolph) [inline src/helpers/folder_helpers.sh]: This is not addressed to me

File: src/git-sync [hunk 13]
```diff
@@ -46,6 +46,7 @@ function skippable {
 
 
 function steps {
+  echo_if_true "change_directory $(git_root)" "$IN_SUB_FOLDER"
```
Comment (charlierudolph) [inline src/git-sync]: I think this is necessary for every command that syncs the main branch (extract, hack, ship, sync, and sync-fork). Would be nice if the cd was the first and last command (ie ran before preconditions)

File: src/helpers/folder_helpers.sh (hunk 5 above)
Comment (kevgo) [inline src/helpers/folder_helpers.sh]: I can't use `run_command` in this situation, because I'd like for it to display the Git branch here, so that this step can be listed and verified in the subsequent `Then it runs the Git commands` step.

This has several benefits:
* it is important on which branch the `cd` operation happens, since the folder cd'ed out of only exists on a particular branch
* when the `cd` command happens in the order of steps is important, because if it happens at the wrong time it leaves the file system in limbo
* it allows to document when exactly we `cd` back into this folder at the end

This means the `Then it runs the Git commands` step probably has to be renamed to `Then it runs the following commands in my workspace` to account for the fact that we now also run non-Git commands in the current workspace. 

Since this is a larger operation, I'd like to do it separately. 

I'm open to doing this differently, but I don't see a way to verify the `cd` command in the spec in another similarly clean way right now. If you see one, please outline it.

File: src/git-sync (hunk 13 above)
Comment (kevgo) [inline src/git-sync]: Agreed. I'll add it to the other commands in separate PRs. I have added checkboxes to #371 to track when all commands are addressed.

**Comment by kevgo** [comment]
I added the `git sync --all` spec to document that the `cd` really only happens at the beginning and end, and that all the branches are sync'ed between it. But I get your point about this being redundant. I am therefore keeping only one spec. The `git sync --all` spec is more comprehensive and expressive, so I keep that one, and removed the `git sync` one. I also moved it to a higher-level folder, since this spec applies to all variations of git-sync.

**Comment by kevgo** [comment]
All comments addressed

File: src/helpers/folder_helpers.sh (hunk 5 above)
Comment (charlierudolph) [inline src/helpers/folder_helpers.sh]: > I can't use `run_command` in this situation, because I'd like for it to display the Git branch here, so that this step can be listed and verified in the subsequent `Then it runs the Git commands` step.

I suggested changing the step that verifies which commands are run to make the branch optional. I don't think the `cd` should display the branch. If you agree, maybe we would want the branch name in the table to say `<none>` (I don't think the branch name being empty would work since we are drying up the branch names).

> it is important on which branch the cd operation happens, since the folder cd'ed out of only exists on a particular branch

This is true for the feature you wrote, but is not guaranteed to always be true.

Comment (kevgo) [inline src/helpers/folder_helpers.sh]: I have updated #493 to reflect the outcome of our conversation. This will be addressed separately to keep this PR focussed.

**Comment by kevgo** [comment]
Ready to ship?

File: src/helpers/helpers.sh [hunk 14]
```diff
@@ -36,7 +37,12 @@ source "$current_dir/environment.sh" "$@"
 source "$current_dir/configuration.sh" "$@"
 
 if [ "$(is_git_repository)" == true ]; then
+  temp_filename_suffix="$(git_root | tr '/' '_')"
+  export STEPS_FILE="/tmp/${PROGRAM}_${temp_filename_suffix}"
+  export UNDO_STEPS_FILE="/tmp/${PROGRAM}_undo_${temp_filename_suffix}"
   export INITIAL_BRANCH_NAME=$(get_current_branch_name)
+  export INITIAL_DIRECTORY=$(pwd)
+  export IN_SUB_FOLDER=$(is_in_git_sub_directory)
```
Comment (charlierudolph) [inline src/helpers/helpers.sh]: This should either be before all the `INITIAL_...` or after. Based on ascii it appears it should go after
```
$ irb
2.2.0 :001 > "_" < "I"
 => false 
2.2.0 :002 > "I" < "_"
 => true 
2.2.0 :003 > 

```

**Comment by charlierudolph** [comment]
:ship: after final comment is addressed
